src/tools/cephfs_mirror/Mirror.cc

```diff
@@ -319,10 +319,11 @@ void Mirror::handle_enable_mirroring(const Filesystem &filesystem,
 
   std::scoped_lock locker(m_lock);
   auto &mirror_action = m_mirror_actions.at(filesystem);
-  ceph_assert(mirror_action.action_in_progress);
-
-  mirror_action.action_in_progress = false;
-  m_cond.notify_all();
+  if (!mirror_action.fs_mirror->is_init_failed()) {
```

Comment:
If this get executed by more than one C_RestartMirroring context completions, it can have the sam effect of hitting the assert.

Reply:
in that case where init() passed then `mirror_action.action_in_progress` will be true and the assertion would pass. That should not be an issue i think

Reply:
Yeh, restart isn't invoked if init doesn't fail.

Reply:
so even if there are multiple threads running into this condition, if init failed then skip the entire block else if passed, `action_in_progress` is anyways true and thus no issue in asserting it.

Reply:
That kind of works, but I feel the correct solution is to disallow running concurrent C_RestartMirroring contexts.

Reply:
concurrent C_RestartMirroring i.e. multiple threads trying to modify `action_in_progress` of the same fs which is definitely incorrect and im support of fixing the root cause but there can be instances in future where something failed in the daemon and we are here again at this assertion failure. This patch might not fix the root issue but I think this code is a good exception handler whenever there is an init failure and this same block exists for disable_mirroring, so this code in enable_mirroring is a must have.

Reply:
> concurrent C_RestartMirroring i.e. multiple threads trying to modify `action_in_progress` of the same fs

minor correction - there aren't multiple threads, just _a_ workqueue driving async context completion.

> which is definitely incorrect and im support of fixing the root cause but there can be instances in future where something failed in the daemon and we are here again at this assertion failure.

The _something_ failure is going to trigger a C_RestartMirroring async task, so, if multiple async context completions are not triggered, then this assert is not to be hit.

> This patch might not fix the root issue but I think this code is a good exception handler whenever there is an init failure and this same block exists for disable_mirroring, so this code in enable_mirroring is a must have.

The error handling block exists in `handle_disable_mirroring` for a reason - if enabling mirroring failed (init failure), then calling 

```
mirror_action.fs_mirror->shutdown(new C_AsyncCallback<ContextWQ>(m_work_queue, on_finish));
```

is going to cause problems, so, disable_mirroring handles init failures by completing the context and therefore the same handling needs to be done in handle_disable_mirorring.

Reply:
>The something failure is going to trigger a C_RestartMirroring async task, so, if multiple async context completions are not triggered, then this assert is not to be hit.

For now yes, there could be significant code changes in future and there could be possibility of multiple triggers, also logically it doesn't make sense to perform any action if the daemon couldn't initialise imo.

Reply:
like this https://github.com/ceph/ceph/pull/52524/commits/5d1c278b25b696d42618878f6c4f7ff5917941be